---
title: 'Metage2Metabo PostaViz: a Python package for exploring, visualising, and comparing the metabolic potential of microbial communities'
tags:
  - Python
  - Systems biology
  - Mmicrobial communities
  - Metabolic modeling
  - Microbiome
  - Visualisation
authors:
  - name: Léonard Brindel
    affiliation: 1
  - name: Clémence Frioux
    orcid: 0000-0003-2114-0697
    corresponding: true
    affiliation: 1
affiliations:
 - name: Inria, Univ. Bordeaux, INRAE, F-33400 Talence, France
   index: 1
date: 31 January 2025
bibliography: paper.bib
---

# Summary

Microbial communities consist of up to thousands of distinct microbial populations, each characterized by its genomic DNA sequences, and all sharing a habitat and environmental conditions. The word _microbiome_ describe the holistic concept associated all the previous components [@Berg.2020]. Characterizing the populations in samples and understanding both their roles of and interactions within the microbiome require a combination of experimentation, high-throughput data acquisition and computational models [@Klitgord.2011]. The role of microbial populations can be abstracted by the study of their metabolism, encompassing all biochemical reactions that cells perform. The genome encodes the genetic information associated to the metabolism, and can therefore be queried to estimate the metabolic capabilities of microorganisms, represented as metabolic networks. Because such metabolism is highly redundant and because of the size of microbial communities, computational models are needed and predict, from the networks, the possible behaviours of and interactions within microbial populations in given environmental conditions. Comparing the outcomes of such predictive models in multiple samples further increases the difficulty of integrating results into actionable hypotheses. Visualisation and integration of metadata can help perform such comparison. They need to be as customisable as possible to facilitate exploration by end-users.  

# Statement of need

Metabolic modelling is widely used to determine the roles of microorganisms in microbial communities [@Cerk.2024]: what molecules they can produce, which interactions are likely to happen depending on the environmental conditions... Many models rely on integer linear programming optimisations [@García-Jimenez.2021] and some alternatives use Boolean abstractions [@Frioux.2018] [@Belcour.2020] or probabilistic approaches [@Bernstein.2019] to provide predictions on community behaviours. Key questions are scalability to large communities, and also integration of these predictions when analysing a large number of samples or community compositions. Each microorganism is abstracted by a collection of biochemical reactions it may perform according to its genomic information, thus forming a network connecting transformed molecules. A community of microorganisms is therefore represented as a collection of such networks, referred to as genome-scale metabolic networks. A metabolic model will provide predictions on the possible behaviour of microorganisms and communities in defined simulation conditions.

`Metage2Metabo-PostaViz` (M2M-PostAViz) is a Python package that performs analyses on the predictions generated by the metabolic-modelling tool `Metage2Metabo` (M2M) [@Belcour.2020]. M2M screens the metabolic potential of a microbial community represented as a collection of genome-scale metabolic networks. When working with cohorts of hundreds or thousands of samples, one has to run the tool as many times as there are samples, then analyze the results of the model. The tool's output are, for each community, several data frames describing the role of each microorganism with respect to the whole community's functions. Properly comparing all samples requires combining all the outputs, and taking into account sample metadata describing individuals lifestyle or clinical characteristics for instance.

`M2M-PostaViz` integrates all such data and provides a visualisation interface that permits exploration through custom plot generation and statistical tests. The underlying data treatment was optimized in order to deal with large numbers of samples without impeding user experience. `M2M-PostaViz` notably permits a pre-treatment and storage of the data such that future exploration can be launched in a computationally efficient manner. Exploration is performed at several levels: molecules (metabolites) that may or may not be producible across samples, microorganisms that may have different behaviours across samples depending on interactions with other community members, or more general overviews of the community functions.

# Outlook

## Data integration and storage

Each sample has to be run in M2M prior analysis. Inputs to the application are a set of directories generated by M2M for each sample, composed of several tabulated and json files. Additional inputs include relative abundance of microbes in each sample, used to weigh the predictions, metadata associated to the samples and possibly metadata related to molecule descriptions and microbial taxonomy.

Reading and integrating all data is computationally demanding when considering several hundreds or several thousands of samples (and M2M outputs directories). Taking into account that users are likely to explore the same data across several runs of the applications, efforts were done to efficiently store the required and processed data such that only the first run takes time and the future ones directly load pre-processed data. The parquet format is used to store all this information... [CF: continue?]

## Application content

The application opens as a multi-tab browser page where the first one is an overview of the data that summarises it and enables a few first analyses customizable by variables of metadata. Two tabs provide analyses centered of microbe roles and molecules (see below) and a last one summarises the metadata and permits customizing variable types to fine-tune analyses.

### Exploration of microorganism roles

The second tab of the application focuses on the role of microorganisms in the productions of metabolites across samples. The same species, and thus the same metabolic network can appear in several samples, but behave differently because of interactions with other microbial populations. In addition, microbial species can be grouped according to their taxonomy, enabling to consider not only the role of a metabolic network but the role of all those falling in taxonomic groups of different levels (phylum, family, genera...). Again, sample metadata variables can refine the analysis, enabling users to compare groups or filter certain samples. 

### Exploration of metabolite production across samples

The third tab of the application centers the analysis on the metabolites, that can also be groupes of families if a proper ontology is provided. Focusing on metabolites enables a targeted analysis in order to compare samples or groups of samples on specific metabolic functions. 

# Acknowledgements

CF is supported by the French National Research Agency (ANR) France 2030 PEPR Agroécologie et Numérique MISTIC ANR-22-PEAE-0011.
Experiments and development were carried on the PlaFRIM experimental testbed, supported by Inria, CNRS (LABRI and IMB), Université de Bordeaux, Bordeaux INP and Conseil Régional d’Aquitaine (see [https://www.plafrim.fr](https://www.plafrim.fr)).

# References
